### PR TITLE
[zero] add zero optimizer for ColoTensor

### PR DESCRIPTION
**Pull request by ver217**

### CODE REVIEW

File: colossalai/nn/optimizer/zero_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+from colossalai.context import ParallelMode
+from colossalai.logging import get_dist_logger
+
+
+class OptimState(Enum):
+    SCALED = 0
+    UNSCALED = 1
+
+
+class ZeroOptimizer(ColossalaiOptimizer):
```
Comment (FrankLeeeee) [inline colossalai/nn/optimizer/zero_optimizer.py:19]: Should this be put in the zero module?

File: colossalai/zero/zero_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        self.grad_scaler = DynamicGradScaler(initial_scale=initial_scale,
+                                             min_scale=min_scale,
+                                             growth_factor=growth_factor,
+                                             backoff_factor=backoff_factor,
+                                             growth_interval=growth_interval,
+                                             hysteresis=hysteresis,
+                                             max_scale=max_scale)
+        self._found_overflow: torch.Tensor = torch.zeros(1, dtype=torch.int64, device=torch.cuda.current_device())
+        self.dp_process_group = gpc.get_group(ParallelMode.DATA)
+        self.mp_process_group = gpc.get_group(ParallelMode.MODEL)
```
[the remark this answers is not in the document]
Comment (ver217) [reply colossalai/zero/zero_optimizer.py:49]: @FrankLeeeee shall we add this mode?

File: colossalai/zero/zero_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                    p.data = self.fp16_param_to_fp32_param[p]
+                else:
+                    assert p.grad is None
+
+    def _update_fp16_params(self):
+        for group in self.optim.param_groups:
+            for p in group['params']:
+                if not self.module.chunk_manager.is_chunk_free(p):
+                    fp32_p = self.fp16_param_to_fp32_param[p]
+                    self.module.chunk_manager.copy_tensor_to_chunk_slice(p, fp32_p)
```
[the remark this answers is not in the document]
Comment (ver217) [reply colossalai/zero/zero_optimizer.py:65]: ok

**Review by feifeibear — APPROVED** [review]

File: colossalai/zero/zero_optimizer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (FrankLeeeee) [reply colossalai/zero/zero_optimizer.py:49]: We have this mode actually...

File: tests/test_tensor/test_gpt.py [hunk 4]
```diff
@@ -68,6 +68,7 @@ def run_gpt(init_spec_func, use_ddp):
     for i, (input_ids, attn_mask) in enumerate(train_dataloader):
         logits = model(input_ids, attn_mask)
         torch_logits = torch_model(input_ids, attn_mask)
+        print(torch_logits, logits)
```
Comment (FrankLeeeee) [inline tests/test_tensor/test_gpt.py:71]: remove this.

**Pull request merged**